backend/metrics-api.py

```diff
@@ -43,6 +47,96 @@ def make_error_response(params, error, status):
+    first step in letting users download arrival data
+    based on the filters they have selected in the frontend
+    '''
+
+    args = request.args
+
+    print(f"args.get('variables'): {args.get('variables')}")
+    variables_dict = json.loads(args.get('variables'))
+
+    agency_id = variables_dict.get('agencyId')
+    agency_config = config.get_agency(agency_id)
+    route_id =variables_dict.get('routeId')
+    direction_id = variables_dict.get('directionId', None)
+    starttime = variables_dict.get('startTime', None)
+    endtime = variables_dict.get('endTime', None)
+
+    arrival_df = pd.DataFrame()
+
+    for date_str in variables_dict.get('dates'):
+
+        date_of_interest = datetime.datetime.strptime(date_str,'%Y-%m-%d').date()
+
+        if (starttime is not None and endtime is not None):
+            starttime_param = util.get_timestamp_or_none(date_of_interest, starttime, agency_config.tz)
+            endtime_param = util.get_timestamp_or_none(date_of_interest, endtime, agency_config.tz)
+        else:
+            starttime_param = None
+            endtime_param = None
+
+        history = arrival_history.get_by_date(agency_id=agency_id
+                                            , route_id=route_id
+                                            , d=date_of_interest
+                                            , version=arrival_history.DefaultVersion)
+        
+        raw_arrival_df = history.get_data_frame(direction_id=direction_id
+                                                ,start_time=starttime_param
+                                                ,end_time=endtime_param
+                                                )
+
+        #raw_arrival_df field names:("VID", "TIME", "DEPARTURE_TIME", "SID", "DID", "DIST", "TRIP")
+        rename_dict = {'TIME':'arrival_time_unix','SID':'stop_id'
+                        ,'DEPARTURE_TIME':'departure_time_unix'
+                        ,'DIST':'calc_gps_distance_to_stop','TRIP':'trip_id'
+                        ,'DID':'direction_id', 'VID':'vehicle_id'}
+
+        partial_arrival_df = raw_arrival_df.rename(columns=rename_dict).copy()
+        partial_arrival_df['date'] = date_of_interest.strftime('%Y-%m-%d')
+
+        if arrival_df.empty:
+            arrival_df = partial_arrival_df
+        else:
+            arrival_df = arrival_df.append(partial_arrival_df)
+
+    # add fields for user context
+    arrival_df['route_id'] = route_id
+    arrival_df['agency'] = agency_id
+        
+
+    ## convert unix timestamp to datetime then convert to agency timezone then format as string
+    arrival_df['arrival_time'] = arrival_df['arrival_time_unix'].apply(lambda x: pytz.utc.localize(datetime.datetime.fromtimestamp(x)).astimezone(agency_config.tz).strftime('%Y-%m-%d %H:%M:%S'))
+    arrival_df['departure_time'] = arrival_df['departure_time_unix'].apply(lambda x: pytz.utc.localize(datetime.datetime.fromtimestamp(x)).astimezone(agency_config.tz).strftime('%Y-%m-%d %H:%M:%S'))
```

Comment:
Save lambda function as variable to avoid duplicating it twice

Reply:
I created a function in util and imported it. Is that what you mean?